File: dbusmenu4j/src/main/java/io/gitlab/jfronny/dbusmenu4j/SwingMenuHolder.java
package io.gitlab.jfronny.dbusmenu4j;

import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

public class SwingMenuHolder<T extends SwingRootMenu<T, ?>> implements MenuHolder {
    protected T root;

    protected SwingMenuHolder(T root) {
        this.root = root;
    }

    private SwingMenuHolder(JMenuBar bar, String menuItem, DMLog log) {
        this.root = (T) new SwingRootMenu<>(bar, menuItem, this, log);
        this.root.syncChildren();
    }

    public static SwingMenuHolder<?> get(JMenuBar bar, String menuItem, DMLog log) {
        return new SwingMenuHolder<>(bar, menuItem, log);
    }

    @Override
    public @Nullable Menu find(int menuId) {
        if (menuId == 0) return root;
        return find(root, menuId);
    }

    private @Nullable Menu find(Menu parent, int menuId) {
        List<Menu> children = parent.getChildren();
        if (children == null) return null;
        for (Menu child : children) {
            if (child.getId() == menuId) return child;
            Menu found = find(child, menuId);
            if (found != null) return found;
        }
        return null;
    }

    public int getId(@Nullable JMenuItem menuItem) {
        return System.identityHashCode(menuItem);
    }
}

File: src/main/kotlin/io/gitlab/jfronny/globalmenu/proxy/SwingMenuHolder.kt
package io.gitlab.jfronny.globalmenu.proxy

import com.intellij.openapi.actionSystem.impl.ActionMenu
import com.intellij.openapi.actionSystem.impl.ActionMenuItem
import io.gitlab.jfronny.dbusmenu4j.Menu
import io.gitlab.jfronny.dbusmenu4j.MenuHolder
import javax.swing.JMenuBar
import javax.swing.JMenuItem

class SwingMenuHolder(bar: JMenuBar, menuName: String): MenuHolder {
    private val root = SwingRootMenu((0 until bar.menuCount).map { bar.getMenu(it) }, menuName, this)

    override fun find(menuId: Int): Menu? {
        if (menuId == 0) return root
        return find(root, menuId)
    }

    private fun find(parent: Menu, menuId: Int): Menu? {
        return parent.children?.let { children ->
            for (child in children) {
                if (child.id == menuId) return child
                val found = find(child, menuId)
                if (found != null) return found
            }
            null
        }
    }

    fun getId(menuItem: JMenuItem?): Int = when (menuItem) {
        is ActionMenu -> menuItem.anAction.toString().hashCode()
        is ActionMenuItem -> menuItem.anAction.toString().hashCode()
        else -> System.identityHashCode(menuItem)
    }
    fun update(items: List<ActionMenu>) = root.update(items)
}